README.md
[![Pub Version](https://img.shields.io/pub/v/verification_code_builder?color=green)](https://pub.dev/packages/verification_code_builder)

# verification_code_builder

Infinitely customizable widget for creating code input forms.

Inspired by [flutter_verification_code](https://pub.dev/packages/flutter_verification_code)

## Features
* Ability to define cell widgets using the builder callback. Only you have complete control over the behavior and look of your cells!
* Built in paste action support.
* Backspace press tracking. Especially useful if you don't like the extra "clear all" button.
* Automatic focus control. You don't need to define the focus switching logic yourself. All you need to do in your cell widgets is to call the `focus.nextFocus()` when the input to the current field is complete - verification_code_builder will take care of the rest.

## WIP
1. Support for derived number of characters in cells
2. Cells RegExp and `onError` callback support
3. `isEditing` flag
4. `getCurrentIndex` method
5. Logo

## Usage
1) Import the library:
```dart
import 'package:verification_code_builder/verification_code_builder.dart';
```

2) Use `VerificationCode` wherever you want:
```dart
@override
Widget build(BuildContext context) {
  return Scaffold(
    appBar: AppBar(
      title: Text('VerificationCode'),
    ),
    body: Center(
      child: VerificationCode(
        length: 6, // Set the number of cells
        onCompleted: (String value) {
          print("Code is $value"); // Set a callback for the result
        },
        builder: (index, focus, textEditingController, pasteMethod) { // Construct your cell widget
          return _generateSingleDigit(index, focus, textEditingController, pasteMethod);
        },
      ),
    )
  );
}

Widget _generateSingleDigit(
  int index, // Cell index in VerificationCode widget
  FocusNode focus, // Focus for "this" cell. VerificationCode use it to manage focus of cells
  TextEditingController textEditingController, // Cell input handler
  PasteMethod pasteMethod // Paste mechanism. Use it for perform paste (e.g. using SelectionControls)
) {
  return Container(
    height: 48,
    width: 48,
    margin: EdgeInsets.only(left: 3),
    decoration: BoxDecoration(
      borderRadius: BorderRadius.all(Radius.circular(8)),
      color: Colors.grey),
    child: Center(
      child: TextField( // If you build your cell based on TextField...
        controller: textEditingController, // just insert textEditingController parameter here
        focusNode: focus, // and focus here
        autofocus: index == 0, // You can use index to perform autofocus on first cell
        // Use this custom SelectionControls widget for easily "native-like" paste experience
        selectionControls: VerificationSelectionControls.create(pasteMethod),
        inputFormatters: [FilteringTextInputFormatter.digitsOnly, LengthLimitingTextInputFormatter(1)],
        keyboardType: TextInputType.number,
        textAlign: TextAlign.center,
        style: TextStyle(
          color: Colors.black,
          fontSize: 30,
          fontWeight: FontWeight.w700,
        ),
        onChanged: (str) => {
          if (str.isNotEmpty) {
            index == 5 ? FocusScope.of(context).unfocus() : focus.nextFocus()
          }
        },
      ),
    ),
  );
}
```

## Showcase
<img src=".github/showcase1.gif" alt="Showcase" width="300">

## Contributing
I will be glad to hear about your successes, ideas and problems that you have encountered while using this widget. You can contact me on [Twitter] and share your experience. I will also be very happy if you send me a small video or gif on which you will show what form for entering the code you got. If you find a bug, please describe it (either on [GitHub], or on [Twitter]). Pull requests with bug fixing and improvements are strongly encouraged and I will be happy to leave a mention of your contribution in the README. I will also be glad to any spelling corrections :)

[Twitter]: https://twitter.com/RareScrap_ru
[GitHub]: https://github.com/RareScrap/verification_code_builder

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.0] - 2021-05-13
### Added
- Migration to null safety.

## [0.0.2] - 2021-05-12
### Added
- `enableHorizontalScroll` flag which allows you to disable horizontal scrolling behaviour.
- Exposed `mainAxisAlignment` field.

## [0.0.1] - 2021-05-10
### Added
- Ability to define cell widgets using the builder callback.
- Paste action support.
- Backspace press tracking.
- Automatic focus control. You don't need to define the focus switching logic yourself. All you need to do in your cell widgets is to call the `focus.nextFocus()` when the input to the current field is complete.
- Example app.

[Unreleased]: https://github.com/RareScrap/verification_code_builder/compare/v0.1.0...HEAD
[0.1.0]: https://github.com/RareScrap/verification_code_builder/releases/tag/v0.1.0
[0.0.2]: https://github.com/RareScrap/verification_code_builder/releases/tag/v0.0.2
[0.0.1]: https://github.com/RareScrap/verification_code_builder/releases/tag/v0.0.1

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:example/example_app.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


test/verification_code_builder_test.dart
// import 'package:flutter_test/flutter_test.dart';

// TODO
void main() {
  // test('adds one to input values', () {
  //   final calculator = Calculator();
  //   expect(calculator.addOne(2), 3);
  //   expect(calculator.addOne(-7), -6);
  //   expect(calculator.addOne(0), 1);
  // });
}


lib/src/verification_code_widget.dart
import 'package:collection/collection.dart' show IterableExtension;
import 'package:flutter/widgets.dart';

import 'backspace_text_editing_controller.dart';

typedef Widget VerificationCodeBuilder(
    int index,
    FocusNode focus,
    TextEditingController textEditingController,
    PasteMethod pasteMethod);
typedef void PasteMethod(String str);

class VerificationCode extends StatefulWidget { // TODO: Stateless?

  final ValueChanged<String> onCompleted;

  final int length;

  final bool enableHorizontalScroll;

  final MainAxisAlignment mainAxisAlignment;

  final VerificationCodeBuilder builder;

  VerificationCode({
    required this.builder,
    required this.onCompleted,
    this.length = 4,
    this.enableHorizontalScroll = true,
    this.mainAxisAlignment = MainAxisAlignment.center,
  });

  @override
  _VerificationCodeState createState() => _VerificationCodeState();
}

class _VerificationCodeState extends State<VerificationCode> {
  final List<FocusNode> _listFocusNode = <FocusNode>[];
  final List<BackspaceTextEditingController> _listControllerText = <BackspaceTextEditingController>[];
  String _code = ""; // To prevent false onComplete firing

  @override
  void initState() {
    for (var i = 0; i < widget.length; i++) {
      _listFocusNode.add(FocusNode());
      var tec = BackspaceTextEditingController(() => {
        if (i > 0) _listFocusNode[i-1].requestFocus()
      });
      // tec.value = new TextEditingValue(text: " ");
      tec.addListener(() {

        // var index = _listControllerText.indexOf(tec);
        // if (tec.text.isEmpty && index > 0) {
        //   tec.value = new TextEditingValue(text: " ");
        //   _listFocusNode[index-2].requestFocus();
        // }


        if (_code != code && isInputComplete) {
          widget.onCompleted(code);
          _code = code;
        }
      });
      _listControllerText.add(tec);
    }

    super.initState();
  }


  @override
  Widget build(BuildContext context) {
    final content = Column(
      children: <Widget>[
        Row(
          mainAxisAlignment: widget.mainAxisAlignment,
          children: _buildListWidget(),
        ),
      ],
    );

    if (widget.enableHorizontalScroll) {
      return SingleChildScrollView(
          scrollDirection: Axis.horizontal,
          child: content);
    } else {
      return content;
    }
  }

  @override
  void dispose() {
    _listFocusNode.forEach((element) { element.dispose(); });
    _listFocusNode.clear();
    _listControllerText.forEach((element) { element.dispose(); });
    _listControllerText.clear();
    super.dispose();
  }

  String get code {
    String verifyCode = ""; // TODO: use accumulation func
    for (var i = 0; i < widget.length; i++) {
      verifyCode += _listControllerText[i].text;
    }
    return verifyCode;
  }

  bool get isInputComplete => _listControllerText.firstWhereOrNull(
          (element) => element.text.isEmpty) == null;

  // TODO: Не учитывает inputFormatters (особенно LengthLimitingTextInputFormatter(2))
  void paste(String str) { // TODO: Не срабатывает при пасте из гугл-клавиатуры
    _listControllerText.forEach((element) { element.enableListeners = false; });

    for (int i = 0; i < widget.length; i++) {
      _listControllerText[i].text = i < str.length ? str[i] : "";
    }
    // Переводим фукус на последнию цифру в вставленной строке
    _listFocusNode[str.length >= widget.length ? widget.length-1 : str.length-1].requestFocus();

    _listControllerText.forEach((element) {
      element.enableListeners = true;
      element.notifyListeners();
    });
  }

  List<Widget> _buildListWidget() {
    List<Widget> listWidget = [];
    for (int index = 0; index < widget.length; index++) {
      // var next = index+1 < _listFocusNode.length ? _listFocusNode[index+1] : null;
      listWidget.add(widget.builder(index, _listFocusNode[index], _listControllerText[index], paste));
    }
    return listWidget;
  }
}

lib/src/backspace_text_editing_controller.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

class BackspaceTextEditingController extends TextEditingController {

  final Function() onLimitBackspacePressed; // TODO: Переименовать и перестать различать крайние бэкспейсы с обычными?

  BackspaceTextEditingController(this.onLimitBackspacePressed);

  bool enableListeners = true;

  @override
  set value(TextEditingValue newValue) {

    // Попытка хандлить softkeyboard backspace, нажатый на пустом текстфилде.
    // Данный метод будет вызван извне при любом изменении текстфилда. Что
    // особенно важно - движок флаттера не проверяет отличается ли новое
    // значение от старого. Именно этим мы и воспользуемся! Логика проста -
    // если новое значение == старому, то вероятнее всего юзер пытается нажать
    // backspace на пустом текстфилде. Однако нужно учитывать пару ньюансов...

    // при попытке двигать ручки селекшона за пределы текстфилда будет каждый фрейм
    // происходить сеттинг нового значения, которое == старому! Флаг ниже проверяет
    // выделен ли текст, чтобы избежать сеттинга одинаковых значений, когда по факту
    // с текстфилдом НИЧЕГО НЕ ПРОИСХОДИТ.
    bool isInSelection = newValue.selection.start != newValue.selection.end;
    // Также проверяем похожую ситуацию, но когда у селекшона только один ханлд
    // bool isMovingSingleSelection = TODO: Сейчас я хз как это проверить

    // Передосим фокус назад, если стерли последний символ
    if (value.text != newValue.text && newValue.text.isEmpty) {
      super.value = const TextEditingValue(selection: TextSelection.collapsed(offset: 0));
      onLimitBackspacePressed();
      return;
    }

    // TODO: Если контент ячейки выделен селекшоном, то новый текст должен заменять контент. Сейчас же он очищается а фокус передается предыдущей ячейке.

    if (value == newValue && value.selection.baseOffset == 0 && !isInSelection) {
      onLimitBackspacePressed();
    } else {
      super.value = newValue;
    }
  }

  @override
  void clearComposing() {
    // clearComposing() имеет свойство вызываться даже когда все и так очищено.
    // Пресекаем подобное поведение.
    var newValue = value.copyWith(composing: TextRange.empty);
    if (newValue != value) value = newValue;
  }

  @override
  void notifyListeners() {
    if (enableListeners) super.notifyListeners();
  }
}

lib/src/verification_code_selection_controls.dart
import 'dart:async';
import 'dart:io';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

import 'verification_code_widget.dart';

class VerificationSelectionControls {
  // TODO: Как правильно чекать платформу?
  static TextSelectionControls create(PasteMethod pasteMethod) =>
      Platform.isIOS
        ? CupertinoVerificationSelectionControls(pasteMethod)
        : MaterialVerificationSelectionControls(pasteMethod);
}

class MaterialVerificationSelectionControls extends MaterialTextSelectionControls {

  final PasteMethod pasteMethod;

  MaterialVerificationSelectionControls(this.pasteMethod);

  @override
  Future<void> handlePaste(TextSelectionDelegate delegate) async {
    final data = await Clipboard.getData(Clipboard.kTextPlain);
    if (data?.text?.isNotEmpty ?? false) pasteMethod(data!.text!);
  }

  @override
  bool canSelectAll(TextSelectionDelegate delegate) => false;

  @override
  bool canCut(TextSelectionDelegate delegate) => false;

  @override
  bool canCopy(TextSelectionDelegate delegate) => false;
}

class CupertinoVerificationSelectionControls extends CupertinoTextSelectionControls {

  final PasteMethod pasteMethod;

  CupertinoVerificationSelectionControls(this.pasteMethod);

  @override
  Future<void> handlePaste(TextSelectionDelegate delegate) async {
    final data = await Clipboard.getData(Clipboard.kTextPlain);
    if (data?.text?.isNotEmpty ?? false) pasteMethod(data!.text!);
  }

  @override
  bool canSelectAll(TextSelectionDelegate delegate) => false;

  @override
  bool canCut(TextSelectionDelegate delegate) => false;

  @override
  bool canCopy(TextSelectionDelegate delegate) => false;
}

example/lib/example_app.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:verification_code_builder/verification_code_builder.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'VerificationCode example app',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text('VerificationCode'),
        ),
        body: Center(
          child: VerificationCode(
            length: 6,
            onCompleted: (String value) {
              print("Code is $value");
            },
            builder: (index, focus, textEditingController, pasteMethod) {
              return _generateSingleDigit(index, focus, textEditingController, pasteMethod);
            },
          ),
        )
    );
  }

  Widget _generateSingleDigit(int i, FocusNode focus, TextEditingController textEditingController, PasteMethod pasteMethod) {
    return Container(
      height: 48,
      width: 48,
      margin: EdgeInsets.only(left: 3),
      decoration: BoxDecoration(
        borderRadius: BorderRadius.all(Radius.circular(8)),
        color: Colors.grey),
      child: Center(
        child: TextField(
          controller: textEditingController,
          focusNode: focus,
          autofocus: i == 0,
          selectionControls: VerificationSelectionControls.create(pasteMethod),
          inputFormatters: [FilteringTextInputFormatter.digitsOnly, LengthLimitingTextInputFormatter(1)],
          keyboardType: TextInputType.number,
          textAlign: TextAlign.center,
          style: TextStyle(
            color: Colors.black,
            fontSize: 30,
            fontWeight: FontWeight.w700,
          ),
          onChanged: (str) => {
            // if (str.isNotEmpty) {
              i == 5 ? FocusScope.of(context).unfocus() : focus.nextFocus() // TODO: Слетает фокус при стирании последней ячейки
            // }
          },
        ),
      ),
    );
  }
}